portal-impl/src/com/liferay/portal/servlet/filters/authverifier/AuthVerifierPipeline.java

```diff
@@ -0,0 +1,573 @@
+				if (_log.isDebugEnabled()) {
+					Class<?> authVerifierClass = authVerifier.getClass();
+
+					if (user == null) {
+						_log.debug(
+							StringBundler.concat(
+								"Auth verifier ", authVerifierClass.getName(),
+								" returned null user",
+								authVerifierResult.getUserId()));
+					}
+					else {
+						_log.debug(
+							StringBundler.concat(
+								"Auth verifier ", authVerifierClass.getName(),
+								" returned inactive user",
+								authVerifierResult.getUserId()));
+					}
+				}
+
+				return null;
+			}
+
+			Map<String, Object> settings = _mergeSettings(
+				properties, authVerifierResult.getSettings());
+
+			settings.put(AUTH_TYPE, authVerifier.getAuthType());
+
+			authVerifierResult.setSettings(settings);
+
+			return authVerifierResult;
+		}
+
+		private final AccessControlContext _accessControlContext;
+		private AuthVerifierResult _authVerifierResult;
+		private Set<AuthVerifierConfiguration>
+			_excludedAuthVerifierConfigurations;
+		private final URLPatternMapper<List<AuthVerifierConfiguration>>
+			_excludeURLPatternMapper;
+		private final String _requestURI;
+
+	}
+
+	private static class AuthVerifierTrackerCustomizer
+		implements ServiceTrackerCustomizer
+			<AuthVerifier, AuthVerifierConfiguration> {
+
+		public static void addAuthVerifierPipeline(
+			AuthVerifierPipeline authVerifierPipeline) {
+
+			_authVerifierPipelines.add(authVerifierPipeline);
+
+			authVerifierPipeline._rebuildConfiguration();
+		}
+
+		@Override
+		public AuthVerifierConfiguration addingService(
+			ServiceReference<AuthVerifier> serviceReference) {
+
+			Registry registry = RegistryUtil.getRegistry();
+
+			AuthVerifier authVerifier = registry.getService(serviceReference);
+
+			if (authVerifier == null) {
+				return null;
+			}
+
+			Class<?> authVerifierClass = authVerifier.getClass();
+
+			AuthVerifierConfiguration authVerifierConfiguration =
+				new AuthVerifierConfiguration();
+
+			authVerifierConfiguration.setAuthVerifier(authVerifier);
+			authVerifierConfiguration.setAuthVerifierClassName(
+				authVerifierClass.getName());
+			authVerifierConfiguration.setProperties(
+				_loadProperties(serviceReference, authVerifierClass.getName()));
+
+			if (!_validate(authVerifierConfiguration)) {
+				return null;
+			}
+
+			_authVerifierConfigurations.add(authVerifierConfiguration);
+
+			_rebuildAll();
+
+			return authVerifierConfiguration;
+		}
+
+		@Override
+		public void modifiedService(
+			ServiceReference<AuthVerifier> serviceReference,
+			AuthVerifierConfiguration authVerifierConfiguration) {
+
+			_authVerifierConfigurations.remove(authVerifierConfiguration);
+
+			authVerifierConfiguration.setProperties(
+				_loadProperties(
+					serviceReference,
+					authVerifierConfiguration.getAuthVerifierClassName()));
+
+			if (_validate(authVerifierConfiguration)) {
+				_authVerifierConfigurations.add(authVerifierConfiguration);
+
+				_rebuildAll();
+			}
+		}
```

Comment:
I believe we discussed this but I don't know if we got to any conclusion... 
we are adding and removing from a `CopyOnWriteArrayList`... would it be possible to just modify in place the configuration... it should not be a big deal because it is not a frequent operation, but still, it seems like a waste of arrays :-)